File: contracts/genlayer/MarketCreator.py
# { "Depends": "py-genlayer:1jb45aa8ynh2a9c9xn3b7qqh8sm5q93hwfp7jqmwsfhh8jpz09h6" }

from genlayer import *

import json


class MarketCreator(gl.Contract):
    def __init__(self):
        pass

    @gl.public.write
    def propose(
        self,
        title: str,
        summary: str,
        source_url: str,
        category_hint: str,
        published_at: str,
    ) -> str:
        if not source_url:
            return json.dumps(
                {
                    "suitable": False,
                    "question": "",
                    "category": "ECONOMY",
                    "resolutionCriteria": "Rejected: missing source URL.",
                    "resolutionSourceUrl": "",
                    "resolutionMode": "GENLAYER",
                    "durationDays": 7,
                    "resolvesAtReason": "No source URL was provided.",
                    "initialProbabilityYes": 50,
                    "reasoning": "Rejected because a named source URL is required.",
                },
                sort_keys=True,
            )

        def nondet() -> str:
            response = gl.nondet.web.get(source_url)
            source_body = response.body.decode("utf-8")
            prompt = f"""
You are creating one binary YES/NO prediction market from a news source.

Source URL: {source_url}
Published at: {published_at}
Category hint: {category_hint}
Title: {title}
Summary: {summary}

Source page content:
{source_body}

Rules:
- Reject if no source URL.
- Reject subjective markets.
- Reject if no clear future event or measurable claim.
- Reject if resolution depends on rumours, private info, pure opinion, or unclear sources.
- Question must be binary YES/NO.
- Resolution criteria must name the exact source and exact condition.
- durationDays must match the event type.
- initialProbabilityYes must be 0-100.
- If uncertain, use 40-60.
- Never use exactly 50 unless there is truly no useful signal.
- Breaking news: 1-3 days.
- Sports/scheduled match: event end + 1 day.
- Politics/election: 7-30 days.
- Economy/CBN/inflation/FX: 7-45 days.
- Commodities/oil/fuel: 7-30 days.
- Security/conflict: 3-14 days.
- Tech/company/product: 3-21 days.
- If no sensible resolution window exists, suitable=false.

Respond only with valid JSON matching this shape:
{{
  "suitable": boolean,
  "question": string,
  "category": "FOREX" | "POLITICS" | "SPORTS" | "ECONOMY" | "SECURITY" | "COMMODITIES" | "TECH",
  "resolutionCriteria": string,
  "resolutionSourceUrl": string,
  "resolutionMode": "GENLAYER",
  "durationDays": integer,
  "resolvesAtReason": string,
  "initialProbabilityYes": integer,
  "reasoning": string
}}
"""
            result = gl.nondet.exec_prompt(prompt).replace("```json", "").replace("```", "").strip()
            return json.dumps(json.loads(result), sort_keys=True)

        return gl.eq_principle.strict_eq(nondet)

File: src/lib/genlayer/marketCreator.ts
import { z } from "zod";
import { callGenLayer, type GenLayerResult } from "./client";

export const genLayerMarketCreatorInputSchema = z.object({
  title: z.string().min(1),
  summary: z.string().default(""),
  sourceUrl: z.string().url(),
  categoryHint: z.string().optional(),
  publishedAt: z.string().optional(),
});

export const genLayerMarketProposalSchema = z.object({
  suitable: z.boolean(),
  question: z.string(),
  category: z.enum(["FOREX", "POLITICS", "SPORTS", "ECONOMY", "SECURITY", "COMMODITIES", "TECH"]),
  resolutionCriteria: z.string(),
  resolutionSourceUrl: z.string().url(),
  resolutionMode: z.literal("GENLAYER"),
  durationDays: z.number().int(),
  resolvesAtReason: z.string(),
  initialProbabilityYes: z.number().int().min(0).max(100),
  reasoning: z.string(),
});

export type GenLayerMarketCreatorInput = z.infer<typeof genLayerMarketCreatorInputSchema>;
export type GenLayerMarketProposal = z.infer<typeof genLayerMarketProposalSchema>;

export async function requestGenLayerMarketProposal(input: GenLayerMarketCreatorInput): Promise<GenLayerResult<GenLayerMarketProposal>> {
  const parsed = genLayerMarketCreatorInputSchema.safeParse(input);
  if (!parsed.success) {
    return { status: "ERROR", error: parsed.error.message };
  }

  return callGenLayer("MARKET_CREATOR", "genlayer_requestMarketCreation", {
    ...parsed.data,
    rules: [
      "Reject if no source URL.",
      "Reject subjective markets.",
      "Reject if no clear future event or measurable claim.",
      "Reject if resolution depends on rumours, private info, pure opinion, or unclear sources.",
      "Question must be binary YES/NO.",
      "Resolution criteria must name the exact source and exact condition.",
      "durationDays must match the event type.",
      "initialProbabilityYes must be 0-100.",
      "If uncertain, use 40-60.",
      "Never use exactly 50 unless there is truly no useful signal.",
      "Breaking news: 1-3 days.",
      "Sports/scheduled match: event end + 1 day.",
      "Politics/election: 7-30 days.",
      "Economy/CBN/inflation/FX: 7-45 days.",
      "Commodities/oil/fuel: 7-30 days.",
      "Security/conflict: 3-14 days.",
      "Tech/company/product: 3-21 days.",
      "If no sensible resolution window exists, suitable=false.",
    ],
  }, validateGenLayerMarketProposal);
}

export async function checkGenLayerMarketProposal(txHash: string): Promise<GenLayerResult<GenLayerMarketProposal>> {
  if (!txHash) return { status: "ERROR", error: "txHash is required" };
  return callGenLayer("MARKET_CREATOR", "genlayer_checkMarketCreation", { txHash }, validateGenLayerMarketProposal);
}

export function validateGenLayerMarketProposal(value: unknown): GenLayerMarketProposal {
  return genLayerMarketProposalSchema.parse(value);
}

export function clampGenLayerDurationDays(durationDays: number) {
  if (!Number.isFinite(durationDays) || durationDays <= 0) return 7;
  return Math.min(45, Math.max(1, Math.round(durationDays)));
}
